migration/tests.py

```diff
@@ -1,49 +1,124 @@
 
-        self.test_external_tool_tab = ExternalToolTab(
-            id='context_external_tool_99999',
-            label='Test External Tool',
-            tool_id=99999,
-            is_hidden=True,
-            position=30
+    def test_get_results_from_pages_with_limit(self):
+        with self.api.client:
+            results = self.api.get_results_from_pages(f'/accounts/{self.account_id}/courses', page_size=5, limit=2)
+        self.assertTrue(len(results) == 2)
+
+    def test_get_retries_on_http_error(self):
+        request = MagicMock(httpx.Request, autospec=True, url=self.course_url)
+        resp = httpx.Response(
+            status_code=httpx.codes.BAD_GATEWAY,
+            request=request
+        )
+        expected_resp = httpx.Response(
+            status_code=httpx.codes.OK,
+            request=request,
+            text=json.dumps(self.course_data)
         )
 
-    def test_find_tab_by_tool_id_returns_tab(self):
-        tab = CourseManager.find_tab_by_tool_id(99999, [self.test_external_tool_tab])
-        self.assertTrue(isinstance(tab, ExternalToolTab))
+        with patch.object(self.api.client, 'get', autospec=True) as mock_get_call:
+            mock_get_call.side_effect = [resp, expected_resp]
+            with self.api.client:
+                result = self.api.get(self.course_url)
+        self.assertEqual(self.course_data, result.data)
+        self.assertEqual(mock_get_call.call_count, 2)
+
+    def test_get_retries_on_decode_error(self):
+        request = MagicMock(httpx.Request, autospec=True, url=self.course_url)
+        bad_json_resp = httpx.Response(
+            status_code=httpx.codes.OK,
+            request=MagicMock(spec=httpx.Request),
+            text=json.dumps(self.course_data)[:-3],
+        )
+        expected_resp = httpx.Response(
+            status_code=httpx.codes.OK,
+            request=request,
+            text=json.dumps(self.course_data)
+        )
+        with patch.object(self.api.client, 'get', autospec=True) as mock_get_call:
+            mock_get_call.side_effect = [bad_json_resp, expected_resp]
+            with self.api.client:
+                result = self.api.get(self.course_url)
+        self.assertEqual(self.course_data, result.data)
+        self.assertEqual(mock_get_call.call_count, 2)
+
+    def test_put_retries_until_failure(self):
+        request = MagicMock(httpx.Request, autospec=True, url=self.course_url)
+        bad_resp = httpx.Response(
+            status_code=httpx.codes.BAD_GATEWAY,
+            request=request
+        )
+        with patch.object(self.api.client, 'put', autospec=True) as mock_put_call:
+            mock_put_call.side_effect = [bad_resp, bad_resp, bad_resp, bad_resp]
+            with self.api.client:
+                with self.assertRaises(httpx.HTTPStatusError):
+                    self.api.put(self.course_url, params={ "name": "Test Course!" })
+        self.assertEqual(mock_put_call.call_count, 4)
```

Comment:
I could definitely add more tests here, but I'm going to leave it like this for now. I think the retrying bit is pretty covered, but success cases for `get` and `put` could be added.

Reply:
Sounds good.